### PR TITLE
SDIT-782 Index prisoner endpoint should remove prisoner…

### PR DESCRIPTION
**Pull request by steverendell**
…  if they don't exist in NOMIS

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/GlobalSearchService.kt [hunk 1]
```diff
@@ -90,7 +90,7 @@ class GlobalSearchService(
     } ?: GlobalResult.NoMatch
   }
 
-  private fun idMatch(globalSearchCriteria: GlobalSearchCriteria): BoolQueryBuilder? {
+  private fun idMatch(globalSearchCriteria: GlobalSearchCriteria): BoolQueryBuilder {
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/GlobalSearchService.kt:93]: compiler warnings

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerIndexService.kt [hunk 2]
```diff
@@ -71,15 +71,26 @@ class PrisonerIndexService(
...
+      }
+      ?: run {
+        telemetryClient.trackEvent(
+          "POSOffenderNotFoundForIndexing",
+          mapOf("prisonerID" to prisonerId),
+          null,
+        )
+        log.error("POSOffenderNotFoundForIndexing {}", prisonerId)
+        return null
+      }
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerIndexService.kt:93]: This means _any_ sync operation will remove a nonexistent prisoner, not just a manual index update

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerIndexService.kt [hunk 3]
```diff
@@ -71,15 +71,26 @@ class PrisonerIndexService(
...
-        mapOf("prisonerID" to prisonerId),
-        null,
-      )
-      log.error("POSOffenderNotFoundForIndexing {}", prisonerId)
-      return null
     }
+      ?: get(prisonerId)?.run {
+        delete(prisonerId)
+        telemetryClient.trackEvent(
+          "POSOffenderRemovedFromIndex",
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerIndexService.kt:78]: I added a `trackEvent` to the `delete` method, don't think we therefore need this event too.  Although I do prefer your event name, mine was named that way to match the `POSPrisonerCreated` and `POSPrisonerUpdated` events...

**Review by petergphillips — APPROVED** [review]
LGTM

**Pull request merged**